File: src/bumper_bot_py/bumper_bot_py/simple_tf_kinematics.py
import rclpy
from rclpy.node import Node
from tf2_ros.static_transform_broadcaster import StaticTransformBroadcaster
from tf2_ros import TransformBroadcaster, TransformException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from geometry_msgs.msg import TransformStamped
import numpy as np
from bumperbot_msgs.srv import GetTransform


class SimpleTfKinematics(Node):
    def __init__(self):
        super().__init__('simple_tf_kinematics')
        
        # Create a StaticTransformBroadcaster Object
        self.static_tf_broadcaster = StaticTransformBroadcaster(self)

        # Create a TransformBroadcaster Object
        self.dynamic_tf_broadcaster = TransformBroadcaster(self)

        # Create a TransformListener Object
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        # create a TransformStamped object to receive the transform
        self.static_tranform_stamped = TransformStamped()
        self.dynamic_tranform_stamped = TransformStamped()

        self.static_tranform_stamped.header.stamp = self.get_clock().now().to_msg()
        self.static_tranform_stamped.header.frame_id = 'bumperbot_base'
        self.static_tranform_stamped.child_frame_id = 'bumperbot_top'
        # Set the translation and rotation
        self.static_tranform_stamped.transform.translation.x = 0.0
        self.static_tranform_stamped.transform.translation.y = 0.0  
        self.static_tranform_stamped.transform.translation.z = 0.3
        self.static_tranform_stamped.transform.rotation.x = 0.0
        self.static_tranform_stamped.transform.rotation.y = 0.0
        self.static_tranform_stamped.transform.rotation.z = 0.0
        self.static_tranform_stamped.transform.rotation.w = 1.0
        

        self.static_tf_broadcaster.sendTransform(self.static_tranform_stamped)
        self.get_logger().info('Static transform broadcaster started')


        self.timer_ = self.create_timer(0.1, self.timer_callback)
        self.x_increment_ = 0.05
        self.x_last_ = 0.0

        self.w_increment_ = np.pi / 180.0
        self.w_last_ = 0.0


        self.get_transform_service = self.create_service(GetTransform, 'get_transform', self.get_transform_callback)



            #method names timer_callback, this method will be called every 1 second
    def get_transform_callback(self, request, response):
        self.get_logger().info('Incoming request: %s and %s' % (request.frame_id, request.child_frame_id))
        requested_transform = TransformStamped()
        try:
            requested_transform  = self.tf_buffer.lookup_transform(request.frame_id, request.child_frame_id, rclpy.time.Time())
        except TransformException as e:
            self.get_logger().error('Transform exception error on : %s and %s ' % (request.frame_id, request.child_frame_id))
            response.transform = None
            response.success = False
            return response
        
        response.transform = requested_transform
        response.success = True
        self.get_logger().info('Outgoing response: %s' % (response.transform))
        return response


    def timer_callback(self):

        # Set the header stamp
        self.dynamic_tranform_stamped.header.stamp = self.get_clock().now().to_msg()
        self.dynamic_tranform_stamped.header.frame_id = 'odom'
        self.dynamic_tranform_stamped.child_frame_id = 'bumperbot_base'

        # Set the translation and rotation
        self.dynamic_tranform_stamped.transform.translation.x = self.x_last_ + self.x_increment_
        self.dynamic_tranform_stamped.transform.translation.y = 0.0  
        self.dynamic_tranform_stamped.transform.translation.z = 0.0
        self.dynamic_tranform_stamped.transform.rotation.x = 0.0
        self.dynamic_tranform_stamped.transform.rotation.y = 0.0
        self.dynamic_tranform_stamped.transform.rotation.z = 0.0
        self.dynamic_tranform_stamped.transform.rotation.w = self.w_last_ + self.w_increment_

        # Send the transform
        self.dynamic_tf_broadcaster.sendTransform(self.dynamic_tranform_stamped)
        self.get_logger().info('Dynamic transform broadcaster started')

        self.x_last_ = self.dynamic_tranform_stamped.transform.translation.x
        self.w_last_ = self.dynamic_tranform_stamped.transform.rotation.w


        


def main(args=None):
    rclpy.init(args=args)
    simple_tf_kinematics = SimpleTfKinematics()
    rclpy.spin(simple_tf_kinematics)
    simple_tf_kinematics.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/bumper_bot_cpp/src/simple_tf_kinematics.cpp
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/static_transform_broadcaster.h"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2_ros/transform_broadcaster.h"



class SimpleTFKinematics : public rclcpp::Node
{
  public:
    SimpleTFKinematics() : Node("simple_tf_kinematics")
    {
      // Create a static transform broadcaster
      static_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
      //std::shared_ptr<tf2::StaticTransformBroadcaster> static_broadcaster_ (new tf2::StaticTransformBroadcaster(this));
      // new is returning a raw pointer of type tf2::StaticTransformBroadcaster which is pointing to SimpleTFKinematics class

      dynamic_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);

      // Create a transform message
      geometry_msgs::msg::TransformStamped static_transform_stamped;


      static_transform_stamped.header.stamp = this->now();  // get_clock()->now();

      static_transform_stamped.header.frame_id = "bumperbot_base";
      static_transform_stamped.child_frame_id = "bumperbot_top";
      static_transform_stamped.transform.translation.x = 0.5;
      static_transform_stamped.transform.translation.y = 0.0;
      static_transform_stamped.transform.translation.z = 0.0;
      static_transform_stamped.transform.rotation.x = 0.0;
      static_transform_stamped.transform.rotation.y = 0.0;
      static_transform_stamped.transform.rotation.z = 0.0;
      static_transform_stamped.transform.rotation.w = 1.0;

      // Send the static transform
      static_broadcaster_->sendTransform(static_transform_stamped);
      timer_ = this->create_wall_timer(std::chrono::milliseconds(100), std::bind(&SimpleTFKinematics::sendDynamicTransform, this));
      
    }

    private:
        std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_broadcaster_;
        std::shared_ptr<tf2_ros::TransformBroadcaster> dynamic_broadcaster_;

        float x_last_ = 0.0;
        float x_increment_ = 0.1;

        rclcpp::TimerBase::SharedPtr timer_;
        geometry_msgs::msg::TransformStamped dynamic_transform_stamped;

        void sendDynamicTransform()
        {
            // Create a transform message
            dynamic_transform_stamped.header.stamp = this->now();  // get_clock()->now();

            dynamic_transform_stamped.header.frame_id = "odom";
            dynamic_transform_stamped.child_frame_id = "bumperbot_base";
            dynamic_transform_stamped.transform.translation.x = x_last_ + x_increment_;
            dynamic_transform_stamped.transform.translation.y = 0.0;
            dynamic_transform_stamped.transform.translation.z = 0.0;
            dynamic_transform_stamped.transform.rotation.x = 0.0;
            dynamic_transform_stamped.transform.rotation.y = 0.0;
            dynamic_transform_stamped.transform.rotation.z = 0.0;
            dynamic_transform_stamped.transform.rotation.w = 1.0;

            // Send the static transform
            dynamic_broadcaster_->sendTransform(dynamic_transform_stamped);

            x_last_= dynamic_transform_stamped.transform.translation.x;

        }
    };


int main (int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleTFKinematics>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}